flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManager.java

```diff
@@ -94,26 +130,218 @@ public RegistrationResponse apply(final JobMasterGateway jobMasterGateway) {
-			String taskExecutorAddress,
-			ResourceID resourceID) {
+	public Future<org.apache.flink.runtime.rpc.registration.RegistrationResponse> registerTaskExecutor(
+		final UUID resourceManagerLeaderId,
+		final String taskExecutorAddress,
+		final ResourceID resourceID
+	) {
+		log.info("received register from taskExecutor {}, address {}", resourceID, taskExecutorAddress);
+		Future<TaskExecutorGateway> taskExecutorFuture =
+			getRpcService().connect(taskExecutorAddress, TaskExecutorGateway.class);
+
+		return taskExecutorFuture.map(new Mapper<TaskExecutorGateway, org.apache.flink.runtime.rpc.registration.RegistrationResponse>() {
+			@Override
+			public org.apache.flink.runtime.rpc.registration.RegistrationResponse apply(final TaskExecutorGateway taskExecutorGateway) {
+				// decline registration if resourceManager cannot connect to the taskExecutor using the given address
+				if(taskExecutorGateway == null) {
+					log.warn("resourceManager {} decline registration from the taskExecutor {}, cannot connect to it using given address {} ",
+						getAddress(), resourceID, taskExecutorAddress);
+					return new org.apache.flink.runtime.rpc.registration.RegistrationResponse.Decline("cannot connect to taskExecutor using given address");
+				} else {
+					// save the register taskExecutor gateway
+					taskExecutorGateways.put(resourceID, taskExecutorGateway);
+					// schedule the heartbeat with the registered taskExecutor
+					ResourceManagerToTaskExecutorHeartbeatScheduler heartbeatScheduler = new ResourceManagerToTaskExecutorHeartbeatScheduler(
+						ResourceManager.this, leaderSessionID, taskExecutorGateway, taskExecutorAddress, resourceID, log);
+					heartbeatScheduler.start();
+					heartbeatSchedulers.put(resourceID, heartbeatScheduler);
+					return new TaskExecutorRegistrationSuccess(new InstanceID(), heartbeatScheduler.getHeartbeatInterval());
+				}
+			}
+		}, getMainThreadExecutionContext());
+
+	}
+
+	/**
+	 * notify resource failure to resourceManager, because of two reasons:
+	 * 1. cannot keep heartbeat with taskManager for several times, mark the resource as failed
+	 * 2. in some corner cases, TM will be marked as invalid by cluster manager master(e.g. yarn master), but TM itself does not realize.
+	 *
+	 * @param resourceID identify the taskManager which to stop
+	 */
+	@RpcMethod
+	public void notifyResourceFailure(ResourceID resourceID) {
+		log.warn("receive failure notification of resource {}", resourceID);
+		TaskExecutorGateway taskManager = taskExecutorGateways.get(resourceID);
+		if(taskManager == null) {
+			// ignore command to stop an unregister taskManager
+			log.warn("ignore stop taskManager command because {} is unregistered", resourceID);
+		} else {
+			taskExecutorGateways.remove(resourceID);
+			closeHeartbeatToResourceIfExist(resourceID);
+			// TODO notify slotManager and notify jobMaster,
+			// TODO slotManager.notifyTaskManagerFailure(resourceID);
+			taskManager.shutDown(leaderSessionID);
```

Comment:
I'm not sure whether other components should be allowed to terminate other components. We could notify the component that it has been marked failed and then let it decide what to do.